# Pandas


Pandas is a library built using NumPy specifically for data analysis. You'll be using Pandas heavily for data manipulation, visualisation, building machine learning models, etc. 

There are two main data structures in Pandas - Series and Dataframes. The default way to store data is dataframes, and thus manipulating dataframes quickly is probably the most important skill set for data analysis. 

<img src="pandas_1.png">

*source: https://pandas.pydata.org/pandas-docs/stable/overview.html*

In this section, you will study:
1. The pandas Series (similar to a numpy array)
    * Creating a pandas series
    * Indexing series
2. Dataframes 
    * Creating dataframes from dictionaries
    * Importing CSV data files as pandas dataframes
    * Reading and summarising dataframes
    * Sorting dataframes 

### 1. The Pandas Series 

A series is similar to a 1-D numpy array, and contains scalar values of the same type (numeric, character, datetime etc.). 
A dataframe is simply a table where each column is a pandas series.


#### Creating Pandas Series

Series are one-dimensional array-like structures, though unlike numpy arrays, they often contain non-numeric data (characters, dates, time, booleans etc.)

You can create pandas series from array-like objects using ```pd.Series()```.

In [1]:
import pandas as pd
pd.Series?

In [2]:
# import pandas, pd is an alias
import pandas as pd

# Creating a numeric pandas series
s = pd.Series([2, 4, 5, 6, 9])
print(s)
print(type(s))

0    2
1    4
2    5
3    6
4    9
dtype: int64
<class 'pandas.core.series.Series'>


Note that each element in the Series has an index, and the index starts at 0 as usual.

In [3]:
# creating a series of characters
# notice that the 'dtype' here is 'object'
char_series = pd.Series(['a', 'b', 'af'])
char_series

0     a
1     b
2    af
dtype: object

In [6]:
# creating a series of type datetime
date_series = pd.date_range(start = '11-09-2017', end = '12-12-2017')
date_series
type(date_series)
date_series[1]

Timestamp('2017-11-10 00:00:00', freq='D')

#### Indexing Series

Indexing series is exactly same as 1-D numpy arrays - index starts at 0.

In [4]:
# Indexing pandas series: Same as indexing 1-d numpy arrays or lists
# accessing the fourth element
s[3]

# accessing elements starting index = 2 till the end
s[2:]

2    5
3    6
4    9
dtype: int64

In [5]:
# accessing the second and the fourth elements
# note that s[1, 3] will not work, you need to pass the indices [1, 3] as a list inside the original []
s[[1, 3]]

1    4
3    6
dtype: int64

Usually, you will work with Series only as a part of dataframes. Let's study the basics of dataframes.

### The Pandas Dataframe

Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having an index and columns having meaningful names.

#### Creating dataframes from dictionaries

There are various ways of creating dataframes, such as creating them from dictionaries, JSON objects, reading from txt, CSV files, etc. 

In [6]:
# keys become column names
df = pd.DataFrame({'name': ['Vinay', 'Kushal', 'Aman', 'Saif'], 
                   'age': [22, 25, 24, 28], 
                    'occupation': ['engineer', 'doctor', 'data analyst', 'teacher']})
df

age    name    occupation
0   22   Vinay      engineer
1   25  Kushal        doctor
2   24    Aman  data analyst
3   28    Saif       teacher

#### Importing CSV data files as pandas dataframes 

For the upcoming exercises, we will use a dataset of a retail store having details about the orders placed, customers, product details, sales, profits etc. 



In [17]:
# reading a CSV file as a dataframe
market_df = pd.read_csv("../pandas/market_fact.csv")


In [18]:
market_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Ord_id               8399 non-null   object 
 1   Prod_id              8399 non-null   object 
 2   Ship_id              8399 non-null   object 
 3   Cust_id              8399 non-null   object 
 4   Sales                8399 non-null   float64
 5   Discount             8399 non-null   float64
 6   Order_Quantity       8399 non-null   int64  
 7   Profit               8399 non-null   float64
 8   Shipping_Cost        8399 non-null   float64
 9   Product_Base_Margin  8336 non-null   float64
dtypes: float64(5), int64(1), object(4)
memory usage: 525.0+ KB


Usually, dataframes are imported as CSV files, but sometimes it is more convenient to convert dictionaries 
into dataframes. For e.g. when the raw data is in a JSON format (which is not uncommon), you can easily convert it into a dictionary, and then into a dataframe. 

You will learn how to convert JSON objects to dataframes later.

#### Reading and Summarising Dataframes

After you import a dataframe, you'd want to quickly understand its structure, shape, meanings of rows and columns etc. Further, you may want to look at summary statistics - such as mean, percentiles etc.

In [19]:
# Looking at the top and bottom entries of dataframes
market_df.head(7)

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   
5  Ord_5446   Prod_6  SHP_7608  Cust_1818   164.02      0.03              23   
6    Ord_31  Prod_12    SHP_41    Cust_26    14.76      0.01               5   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38  
5   -47.64           6.15                 0.37  
6     1.32           0.50                 0.36

In [20]:
market_df.tail(10)

Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
8389  Ord_4444  Prod_13  SHP_6192  Cust_1519   159.4100      0.00   
8390  Ord_5435  Prod_16  SHP_7594  Cust_1798   316.9900      0.04   
8391  Ord_5435   Prod_4  SHP_7594  Cust_1798  1991.8985      0.07   
8392  Ord_5384   Prod_9  SHP_7519  Cust_1798   181.5000      0.08   
8393  Ord_5348   Prod_8  SHP_7470  Cust_1798   356.7200      0.07   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity  Profit  Shipping_Cost  Product_Base_Margin  
8389              44   34.68           0.98                 0.52  
8390              47 -276.54           8.37                 0.58  
8391              20   88.36           7.69                 0.58  
8392              43   -6.24           2.50                 0.37  
8393               9   12.61           1.99                 0.44  
8394              28  374.63           7.69                 0.59  
8395              20  -74.03           6.92                 0.37  
8396              39  -70.85           5.35                 0.40  
8397              23  565.34          30.00                 0.62  
8398              47  131.39           4.86                 0.38

Here, each row represents an order placed at a retail store. Notice the index associated with each row - starts at 0 and ends at 8398, implying that there were 8399 orders placed.

In [21]:
# Looking at the datatypes of each column
market_df.info()

# Note that each column is basically a pandas Series of length 8399
# The ID columns are 'objects', i.e. they are being read as characters
# The rest are numeric (floats or int)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Ord_id               8399 non-null   object 
 1   Prod_id              8399 non-null   object 
 2   Ship_id              8399 non-null   object 
 3   Cust_id              8399 non-null   object 
 4   Sales                8399 non-null   float64
 5   Discount             8399 non-null   float64
 6   Order_Quantity       8399 non-null   int64  
 7   Profit               8399 non-null   float64
 8   Shipping_Cost        8399 non-null   float64
 9   Product_Base_Margin  8336 non-null   float64
dtypes: float64(5), int64(1), object(4)
memory usage: 525.0+ KB


In [22]:
# Describe gives you a summary of all the numeric columns in the dataset
market_df.describe()

Sales     Discount  Order_Quantity        Profit  Shipping_Cost  \
count   8399.000000  8399.000000     8399.000000   8399.000000    8399.000000   
mean    1775.878179     0.049671       25.571735    181.184424      12.838557   
std     3585.050525     0.031823       14.481071   1196.653371      17.264052   
min        2.240000     0.000000        1.000000 -14140.700000       0.490000   
25%      143.195000     0.020000       13.000000    -83.315000       3.300000   
50%      449.420000     0.050000       26.000000     -1.500000       6.070000   
75%     1709.320000     0.080000       38.000000    162.750000      13.990000   
max    89061.050000     0.250000       50.000000  27220.690000     164.730000   

       Product_Base_Margin  
count          8336.000000  
mean              0.512513  
std               0.135589  
min               0.350000  
25%               0.380000  
50%               0.520000  
75%               0.590000  
max               0.850000

In [23]:
# Column names
market_df.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [24]:
# The number of rows and columns
market_df.shape

(8399, 10)

In [25]:
# You can extract the values of a dataframe as a numpy array using df.values 
market_df.values

array([['Ord_5446', 'Prod_16', 'SHP_7609', ..., -30.51, 3.6, 0.56],
       ['Ord_5406', 'Prod_13', 'SHP_7549', ..., 4.56, 0.93, 0.54],
       ['Ord_5446', 'Prod_4', 'SHP_7610', ..., 1148.9, 2.5, 0.59],
       ...,
       ['Ord_5388', 'Prod_6', 'SHP_7524', ..., -70.85, 5.35, 0.4],
       ['Ord_5348', 'Prod_15', 'SHP_7469', ..., 565.34, 30.0, 0.62],
       ['Ord_5459', 'Prod_6', 'SHP_7628', ..., 131.39, 4.86, 0.38]],
      dtype=object)

#### Indices 

An important concept in pandas dataframes is that of *row indices*. By default, each row is assigned indices starting from 0, and are represented at the left side of the dataframe. 

In [26]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

Now, arbitrary numeric indices are difficult to read and work with. Thus, you may want to change the indices of the df to something more meanigful.

Let's change the index to Ord_id (unique id of each order), so that you can select rows using the order ids directly.

In [15]:
# Setting index to Ord_id
market_df.set_index('Ord_id',inplace=True)
market_df.head()

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38

Having meaningful row labels as indices helps you to select (subset) dataframes easily. You will study selecting dataframes in the next section. 

#### Sorting dataframes

You can sort dataframes in two ways - 1) by the indices and 2) by the values.  


In [28]:
# Sorting by index
# axis = 0 indicates that you want to sort rows (use axis=1 for columns)
market_df.sort_index(axis = 0, ascending = True)

Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
0     Ord_5446  Prod_16  SHP_7609  Cust_1818   136.8100      0.01   
1     Ord_5406  Prod_13  SHP_7549  Cust_1818    42.2700      0.01   
2     Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.6900      0.00   
3     Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.8900      0.09   
4     Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.1500      0.08   
...        ...      ...       ...        ...        ...       ...   
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity   Profit  Shipping_Cost  Product_Base_Margin  
0                 23   -30.51           3.60                 0.56  
1                 13     4.56           0.93                 0.54  
2                 26  1148.90           2.50                 0.59  
3                 43   729.34          14.30                 0.37  
4                 35  1219.87          26.30                 0.38  
...              ...      ...            ...                  ...  
8394              28   374.63           7.69                 0.59  
8395              20   -74.03           6.92                 0.37  
8396              39   -70.85           5.35                 0.40  
8397              23   565.34          30.00                 0.62  
8398              47   131.39           4.86                 0.38  

[8399 rows x 10 columns]

In [32]:
# Sorting by values

# Sorting in increasing order of Sales
market_df.sort_values(by='Sales').head(10)

Ord_id  Prod_id   Ship_id    Cust_id  Sales  Discount  Order_Quantity  \
6700   Ord_704   Prod_7   SHP_964   Cust_242   2.24      0.01               1   
4631   Ord_149   Prod_3  SHP_7028  Cust_1712   3.20      0.09               1   
817   Ord_4270   Prod_7  SHP_5959  Cust_1450   3.23      0.06               2   
52    Ord_4755  Prod_13  SHP_6628  Cust_1579   3.41      0.06               1   
384   Ord_2252   Prod_3  SHP_3064   Cust_881   3.42      0.05               1   
5166  Ord_1109  Prod_13  SHP_1531   Cust_429   3.63      0.03               1   
4510  Ord_1751  Prod_13  SHP_2427   Cust_597   3.77      0.09               1   
45    Ord_4768  Prod_12  SHP_6650  Cust_1579   3.85      0.08               1   
3874  Ord_2380   Prod_7  SHP_3268   Cust_953   3.96      0.10               3   
2453  Ord_2890   Prod_7  SHP_3987  Cust_1054   4.94      0.10               3   

      Profit  Shipping_Cost  Product_Base_Margin  
6700   -1.97           0.70                 0.37  
4631   -3.16           1.49                 0.37  
817    -2.73           0.70                 0.81  
52     -1.78           0.70                 0.56  
384    -2.91           1.49                 0.37  
5166   -1.56           0.70                 0.56  
4510   -1.84           0.96                 0.58  
45     -1.36           0.99                 0.37  
3874   -1.72           0.70                 0.38  
2453   -1.76           0.70                 0.37

In [33]:
# Sorting in decreasing order of Shipping_Cost
market_df.sort_values(by='Shipping_Cost', ascending = False).head()

Ord_id  Prod_id   Ship_id   Cust_id      Sales  Discount  \
4509  Ord_1751  Prod_15  SHP_2426  Cust_597  14740.510      0.00   
5327   Ord_839  Prod_11  SHP_1361  Cust_364  12689.870      0.04   
8283  Ord_1741  Prod_11  SHP_2411  Cust_595  15168.820      0.02   
2800   Ord_417  Prod_11   SHP_561  Cust_156  20333.816      0.02   
5511  Ord_1581  Prod_15  SHP_2184  Cust_519   2573.920      0.07   

      Order_Quantity   Profit  Shipping_Cost  Product_Base_Margin  
4509              46  3407.73         164.73                 0.56  
5327              44  -169.23         154.12                 0.76  
8283              26 -1096.78         147.12                 0.80  
2800              45 -1430.45         147.12                 0.80  
5511              17   117.23         143.71                 0.55

In [39]:
# Sorting by more than two columns

# Sorting in ascending order of Sales for each Product
market_df.sort_values(by=['Prod_id', 'Sales'], ascending = False).T

Ord_id               Ord_2197   Ord_4356  Ord_262   Ord_4059   Ord_2973  \
Prod_id                Prod_9     Prod_9   Prod_9     Prod_9     Prod_9   
Ship_id              SHP_2994   SHP_6074  SHP_358   SHP_5660   SHP_6073   
Cust_id              Cust_827  Cust_1481  Cust_66  Cust_1378  Cust_1480   
Sales                  7522.8    6831.72  6553.45     5587.2    5410.95   
Discount                 0.04       0.01     0.03       0.05       0.09   
Order_Quantity             48         41       39         36         36   
Profit                3187.37    3081.02  2969.81    2254.16    2077.91   
Shipping_Cost           19.99      19.99    19.99      19.99      19.99   
Product_Base_Margin      0.39       0.39     0.39       0.39       0.39   

Ord_id                Ord_950   Ord_5112   Ord_612   Ord_3443   Ord_3650  ...  \
Prod_id                Prod_9     Prod_9    Prod_9     Prod_9     Prod_9  ...   
Ship_id              SHP_1315   SHP_7141   SHP_836   SHP_4773   SHP_6795  ...   
Cust_id              Cust_334  Cust_1729  Cust_687  Cust_1246  Cust_1683  ...   
Sales                 4906.85    4273.95   3872.38    3849.17    3353.54  ...   
Discount                 0.09       0.05       0.1       0.06       0.07  ...   
Order_Quantity             32         49        50         46         22  ...   
Profit                1907.94    1340.07   1110.35    1982.78    1189.96  ...   
Shipping_Cost           19.99      19.99     19.99       5.01      19.99  ...   
Product_Base_Margin      0.39        0.4      0.38       0.38       0.39  ...   

Ord_id                Ord_4079   Ord_362    Ord_130  Ord_1971  Ord_1830  \
Prod_id                 Prod_1    Prod_1     Prod_1    Prod_1    Prod_1   
Ship_id               SHP_5686   SHP_481   SHP_3971  SHP_3185  SHP_2524   
Cust_id              Cust_1386  Cust_144  Cust_1064  Cust_919  Cust_634   
Sales                    52.59     47.01      40.06      32.6     29.65   
Discount                  0.05      0.09       0.07      0.03      0.05   
Order_Quantity               4         1          6         2         1   
Profit                  -30.39    -19.32     -24.44    -20.57    -16.54   
Shipping_Cost              8.6     19.99       4.69       5.5      7.69   
Product_Base_Margin       0.57      0.67       0.68      0.57      0.57   

Ord_id                Ord_3713   Ord_2746   Ord_439  Ord_2314  Ord_286  
Prod_id                 Prod_1     Prod_1    Prod_1    Prod_1   Prod_1  
Ship_id               SHP_5145   SHP_3767   SHP_587  SHP_3171  SHP_387  
Cust_id              Cust_1307  Cust_1030  Cust_136  Cust_899  Cust_90  
Sales                    27.83      22.61     18.73     18.16    18.15  
Discount                  0.09       0.03      0.05      0.03     0.04  
Order_Quantity               2          1         1         1        1  
Profit                  -22.14       -8.4     -6.68     -7.25    -7.26  
Shipping_Cost             9.45       7.51      6.13      6.13     6.13  
Product_Base_Margin        0.6       0.57      0.57      0.57     0.57  

[9 rows x 8399 columns]